src/rewrites.rs

```diff
@@ -6,7 +6,98 @@
 //! Module that allows to rewrite request URLs with pattern matching support.
 //!
 
-use crate::settings::Rewrites;
+use headers::HeaderValue;
+use hyper::{header::HOST, Body, Request, Response, StatusCode, Uri};
+
+use crate::{
+    handler::RequestHandlerOpts,
+    redirects::{handle_error, replace_placeholders},
+    settings::{file::RedirectsKind, Rewrites},
+    Error,
+};
+
+/// Applies rewrite rules to a request if necessary.
+pub(crate) fn pre_process(
+    opts: &RequestHandlerOpts,
+    req: &mut Request<Body>,
+) -> Option<Result<Response<Body>, Error>> {
+    let rewrites = opts.advanced_opts.as_ref()?.rewrites.as_deref()?;
+    let uri_path = req.uri().path();
+
+    let matched = rewrite_uri_path(uri_path, Some(rewrites))?;
+    let dest = match replace_placeholders(uri_path, &matched.source, &matched.destination) {
+        Ok(dest) => dest,
+        Err(err) => return handle_error(err, opts, req),
+    };
+
+    if let Some(redirect_type) = &matched.redirect {
+        // Handle redirects
+        let loc = match HeaderValue::from_str(&dest) {
+            Ok(val) => val,
+            Err(err) => {
+                return handle_error(
+                    Error::new(err).context("invalid header value from current uri"),
+                    opts,
+                    req,
+                )
+            }
+        };
+        let mut resp = Response::new(Body::empty());
+        resp.headers_mut().insert(hyper::header::LOCATION, loc);
+        *resp.status_mut() = match redirect_type {
+            RedirectsKind::Permanent => StatusCode::MOVED_PERMANENTLY,
+            RedirectsKind::Temporary => StatusCode::FOUND,
+        };
+        Some(Ok(resp))
+    } else {
+        // Handle internal rewrites
+        *req.uri_mut() = match merge_uris(req.uri(), &dest) {
+            Ok(uri) => uri,
+            Err(err) => {
+                return handle_error(
+                    err.context("invalid rewrite target from current uri"),
+                    opts,
+                    req,
+                )
+            }
+        };
+
+        // Adjust Host header to allow rewriting to a different virtual host
+        if let Some(host) = req.uri().host() {
+            let mut host = host.to_owned();
+            if let Some(port) = req.uri().port_u16() {
+                host.push_str(&format!(":{}", port));
+            }
+            if let Ok(host) = host.parse() {
+                req.headers_mut().insert(HOST, host);
+            }
+        }
+
+        None
+    }
```

Comment:
I agree that this new improved behavior has to be documented, even showing an example and mentioning the possible caveats described in the PR.

Reply:
I tried documenting this, not sure how understandable it is.